### ISSUE TITLE
How to do Early stopping with monitoring weighted metric of multi-output model?

### ISSUE THREAD

**Issue by Liu-Da** [tensorflow/tensorflow#35857]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Colab
- TensorFlow version (use command below): 2.1
- Python version: 3

**Describe the current behavior**
Is there any way to do early stopping with monitoring weighted metric of multi-output model?
I want to have a "global" metric like weighted accuracy to control my training process.
But I cant find any example or solution on the internet.  

**Code to reproduce the issue**
https://colab.research.google.com/drive/1mcqYo0OFs91uxo3mN82UpLCqCymSKTAI#scrollTo=vqTq1JH5U7Fj
```
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

image_input = keras.Input(shape=(32, 32, 3), name='img_input')
timeseries_input = keras.Input(shape=(None, 10), name='ts_input')

x1 = layers.Conv2D(3, 3)(image_input)
x1 = layers.GlobalMaxPooling2D()(x1)

x2 = layers.Conv1D(3, 3)(timeseries_input)
x2 = layers.GlobalMaxPooling1D()(x2)

x = layers.concatenate([x1, x2])

score_output = layers.Dense(1, name='score_output')(x)
class_output = layers.Dense(5, activation='softmax', name='class_output')(x)

model = keras.Model(inputs=[image_input, timeseries_input],
                    outputs=[score_output, class_output])
model.compile(
    optimizer=keras.optimizers.RMSprop(1e-3),
    loss={'score_output': keras.losses.MeanSquaredError(),
          'class_output': keras.losses.CategoricalCrossentropy()},
    metrics={'score_output': [keras.metrics.CategoricalAccuracy()],
             'class_output': [keras.metrics.CategoricalAccuracy()]},
    loss_weights={'score_output': 2., 'class_output': 1.})
callbacks = [
    # keras.callbacks.EarlyStopping("score_output_categorical_accuracy", patience=2, restore_best_weights=True, mode="auto")
    keras.callbacks.EarlyStopping("weighted_categorical_accuracy", patience=2, restore_best_weights=True, mode="auto")
]
import numpy as np
# Generate dummy Numpy data
img_data = np.random.random_sample(size=(100, 32, 32, 3))
ts_data = np.random.random_sample(size=(100, 20, 10))
score_targets = np.random.random_sample(size=(100, 1))
class_targets = np.random.random_sample(size=(100, 5))

# Fit on lists
model.fit([img_data, ts_data], [score_targets, class_targets],
          callbacks=callbacks,
          batch_size=32,
          epochs=10)
```

**Other info / logs**
`WARNING:tensorflow:Early stopping conditioned on metric `weighted_categorical_accuracy` which is not available.`

**Comment by amahendrakar** [comment]
@Liu-Da,
I do not have the permission to view the colab link you have provided. Please try saving the Gist using the following method 'File' -> 'Save a copy as Github Gist', and share the link of the new window.

Also, I tried to reproduce the issue and did not get any warnings in the code. Please find the Gist [here](https://colab.sandbox.google.com/gist/amahendrakar/15f7010d7bde816ddd1d5e7c33f6edc2/35857.ipynb). Thanks!

**Comment by Liu-Da** [comment]
> @Liu-Da,
> I do not have the permission to view the colab link you have provided. Please try saving the Gist using the following method 'File' -> 'Save a copy as Github Gist', and share the link of the new window.
> 
> Also, I tried to reproduce the issue and did not get any warnings in the code. Please find the Gist [here](https://colab.sandbox.google.com/gist/amahendrakar/15f7010d7bde816ddd1d5e7c33f6edc2/35857.ipynb). Thanks!

@amahendrakar 
Actually, there is warining in your [Gist](https://colab.research.google.com/gist/amahendrakar/15f7010d7bde816ddd1d5e7c33f6edc2/35857.ipynb).
![image](https://user-images.githubusercontent.com/24472521/72412560-97db8180-37a8-11ea-8e5b-4655ed966850.png)
This warning indicate that the early stopping callback cant find **weighted_categorical_accuracy** metric.

**Comment by jvishnuvardhan** [comment]
@Liu-Da did you define any `custom_loss function`? As there is no loss/custom_loss function with name `"weighted_categorical_accuracy"` in your code, it was throwing that `warning`. Please define a custom_loss function with that name then there will be no issue.

There are several ways to write a custom loss function. Here is one way of writing it.
```
import tensorflow as tf
from tensorflow.keras.layers import Dense, Input
from tensorflow.keras.models import Model
from tensorflow.keras.losses import Loss
import numpy as np

inp = Input(shape=(1,), dtype=tf.float32)
targets = Input(shape=(1,), dtype=tf.float32)
w = Input(shape=(1,), dtype=tf.float32)
out = Dense(1)(inp)
m = Model(inputs=[inp, w, targets], outputs=out)



def weighted_loss(y_true, y_pred, weights):
      loss = tf.math.squared_difference(y_pred, y_true)
      w_loss = tf.multiply(weights, loss)
      return tf.reduce_mean(tf.reduce_sum(w_loss, axis=-1))
  
m.add_loss(weighted_loss(targets, out, w))

m.compile(loss=None, optimizer='adam')
x = np.ones((512,1))
m.fit([x, x, x], epochs=10)
m.save('test.h5')      
m2 = tf.keras.models.load_model('test.h5')
m2.fit([x, x, x], epochs=10)
```

I am closing this issue as it was resolved. Feel free to post further questions in Stackoverflow as the focus of GitHub repo is mainly for Bugs and performance related issue. Thanks!

**Comment by Liu-Da** [comment]
> @Liu-Da did you define any `custom_loss function`? As there is no loss/custom_loss function with name `"weighted_categorical_accuracy"` in your code, it was throwing that `warning`. Please define a custom_loss function with that name then there will be no issue.
> 
> There are several ways to write a custom loss function. Here is one way of writing it.
> 
> ```
> import tensorflow as tf
> from tensorflow.keras.layers import Dense, Input
> from tensorflow.keras.models import Model
> from tensorflow.keras.losses import Loss
> import numpy as np
> 
> inp = Input(shape=(1,), dtype=tf.float32)
> targets = Input(shape=(1,), dtype=tf.float32)
> w = Input(shape=(1,), dtype=tf.float32)
> out = Dense(1)(inp)
> m = Model(inputs=[inp, w, targets], outputs=out)
> 
> 
> 
> def weighted_loss(y_true, y_pred, weights):
>       loss = tf.math.squared_difference(y_pred, y_true)
>       w_loss = tf.multiply(weights, loss)
>       return tf.reduce_mean(tf.reduce_sum(w_loss, axis=-1))
>   
> m.add_loss(weighted_loss(targets, out, w))
> 
> m.compile(loss=None, optimizer='adam')
> x = np.ones((512,1))
> m.fit([x, x, x], epochs=10)
> m.save('test.h5')      
> m2 = tf.keras.models.load_model('test.h5')
> m2.fit([x, x, x], epochs=10)
> ```
> 
> I am closing this issue as it was resolved. Feel free to post further questions in Stackoverflow as the focus of GitHub repo is mainly for Bugs and performance related issue. Thanks!

@jvishnuvardhan 
It seems that I did not describe my problem well. Let me explain in more detail. When I define a model with multiple outputs, we can define a built-in metric or a custom metric for each output to describe the performance of the model. But my question is how can I define a weighted metric on multiple output nodes.

for example. If my model has three output `out1` `out2` `out3`, their metrics are `accuracy`, and correspondingly we can get `acc1` `acc2` `acc3`. My question is if I can define a new metric based on three acc (`weighted_acc = (acc1 + acc2 + acc3) / 3`) and use this metric as an indicator for earlystopping monitoring. This indicator does not correspond to an output node, but describes the performance of the entire model.

In my understanding, the `weighted_loss` mentioned in your code is defined for a single output. I couldn't find any information on how to define a metric for multiple outputs. 

I would be very grateful if you could give me any suggestions on this.

**Comment by Liu-Da** [comment]
@jvishnuvardhan 
I uploaded an image to further describe my problem.
![361579404857_ pic_hd](https://user-images.githubusercontent.com/24472521/72674290-eee69c80-3aaf-11ea-9501-03b4da7933f1.jpg)

**Comment by Liu-Da** [comment]
@jvishnuvardhan 
I found a similar issue in the keras project, but there is no feasible solution. https://github.com/keras-team/keras/issues/10287

**Comment by QuickLearner171998** [comment]
> > @Liu-Da did you define any `custom_loss function`? As there is no loss/custom_loss function with name `"weighted_categorical_accuracy"` in your code, it was throwing that `warning`. Please define a custom_loss function with that name then there will be no issue.
> > There are several ways to write a custom loss function. Here is one way of writing it.
> > ```
> > import tensorflow as tf
> > from tensorflow.keras.layers import Dense, Input
> > from tensorflow.keras.models import Model
> > from tensorflow.keras.losses import Loss
> > import numpy as np
> > 
> > inp = Input(shape=(1,), dtype=tf.float32)
> > targets = Input(shape=(1,), dtype=tf.float32)
> > w = Input(shape=(1,), dtype=tf.float32)
> > out = Dense(1)(inp)
> > m = Model(inputs=[inp, w, targets], outputs=out)
> > 
> > 
> > 
> > def weighted_loss(y_true, y_pred, weights):
> >       loss = tf.math.squared_difference(y_pred, y_true)
> >       w_loss = tf.multiply(weights, loss)
> >       return tf.reduce_mean(tf.reduce_sum(w_loss, axis=-1))
> >   
> > m.add_loss(weighted_loss(targets, out, w))
> > 
> > m.compile(loss=None, optimizer='adam')
> > x = np.ones((512,1))
> > m.fit([x, x, x], epochs=10)
> > m.save('test.h5')      
> > m2 = tf.keras.models.load_model('test.h5')
> > m2.fit([x, x, x], epochs=10)
> > ```
> > 
> > 
> > I am closing this issue as it was resolved. Feel free to post further questions in Stackoverflow as the focus of GitHub repo is mainly for Bugs and performance related issue. Thanks!
> 
> @jvishnuvardhan
> It seems that I did not describe my problem well. Let me explain in more detail. When I define a model with multiple outputs, we can define a built-in metric or a custom metric for each output to describe the performance of the model. But my question is how can I define a weighted metric on multiple output nodes.
> 
> for example. If my model has three output `out1` `out2` `out3`, their metrics are `accuracy`, and correspondingly we can get `acc1` `acc2` `acc3`. My question is if I can define a new metric based on three acc (`weighted_acc = (acc1 + acc2 + acc3) / 3`) and use this metric as an indicator for earlystopping monitoring. This indicator does not correspond to an output node, but describes the performance of the entire model.
> 
> In my understanding, the `weighted_loss` mentioned in your code is defined for a single output. I couldn't find any information on how to define a metric for multiple outputs.
> 
> I would be very grateful if you could give me any suggestions on this.

https://github.com/keras-team/keras/issues/10287#issuecomment-636499210
